FurCoNZ/Services/OrderService.cs

```diff
@@ -45,7 +45,79 @@ public Task<DateTimeOffset> ReserveTicketsForPurchaseAsync(IDictionary<int, int>
+        }
+
+        public async Task<Order> GetUserPendingOrderAsync(User user, CancellationToken cancellationToken = default)
+        {
+            return await _db.Orders
+                .Include(o => o.TicketsPurchased)
+                .ThenInclude(t => t.TicketType)
+                .SingleOrDefaultAsync(o => o.OrderedById == user.Id && o.AmountPaidCents == 0, cancellationToken);
+        }
+
+        public async Task AddReceivedFundsForOrder(int orderId, int amountCents, string paymentProvider, string paymentReference, DateTimeOffset when, CancellationToken cancellationToken = default)
+        {
+            var order = await _db.Orders
+                .Include(o => o.Audits)
+                .SingleAsync(o => o.Id == orderId, cancellationToken);
+
+            if (order.Audits.Any(a => a.PaymentProvider == paymentProvider && a.PaymentProviderReference == paymentReference && a.Type == AuditType.Received))
+                throw new InvalidOperationException($"Received funds for order {orderId} has already been applied for {paymentProvider}: {paymentReference}");
+
+            var audit = new OrderAudit
+            {
+                OrderId = orderId,
+                PaymentProvider = paymentProvider,
+                PaymentProviderReference = paymentReference,
+                Type = AuditType.Received,
+                When = when,
+                AmountCents = amountCents,
+            };
+
+            // Recalculate amount paid
+            order.AmountPaidCents = order.Audits.Sum(a => a.AmountCents) + audit.AmountCents;
+
+            await _db.OrderAudits.AddAsync(audit, cancellationToken);
+            await _db.SaveChangesAsync(cancellationToken);
+        }
+
+        public async Task RefundFundsForOrder(int orderId, int amountCents, string paymentProvider, string paymentReference, DateTimeOffset when, CancellationToken cancellationToken = default)
+        {
+            var order = await _db.Orders
+                .Include(o => o.Audits)
+                .SingleAsync(o => o.Id == orderId, cancellationToken);
+
+            if(order.Audits.Any(a => a.PaymentProvider == paymentProvider && a.PaymentProviderReference == paymentReference && a.Type == AuditType.Refunded))
+                throw new InvalidOperationException($"Refund for order {orderId} has already been applied for {paymentProvider}: {paymentReference}");
+            
+
+            var audit = new OrderAudit
+            {
+                OrderId = orderId,
+                PaymentProvider = paymentProvider,
+                PaymentProviderReference = paymentReference,
+                Type = AuditType.Refunded,
+                When = when,
+                AmountCents = -amountCents,
+            };
+
+            // Recalculate amount paid
+            order.AmountPaidCents = order.Audits.Sum(a => a.AmountCents) + audit.AmountCents;
+
+            await _db.OrderAudits.AddAsync(audit, cancellationToken);
```

Comment:
Don't need to use Async here, since Async is only used in scenarios where a value needs to be generated by the database prior to committing the data. Microsoft recommends using `.Add()` in all other cases (see https://github.com/aspnet/EntityFrameworkCore/blob/master/src/EFCore/DbContext.cs#L788-L790).
```suggestion
            _db.OrderAudits.Add(audit);
```